rfcs/0007-handling-corruption.md

```diff
@@ -0,0 +1,468 @@
+application can provide operators with detailed information on the corruption,
+including suggestions on how to fix it.
+
+#### Handling Manifest Corruption during Open
+If corruption is detected in a manifest when calling `open_with_opts()`, the
+callback `DBOptions.on_corruption` is invoked. Regardless of the return value
+of the callback, `open_with_opts()` will return an error, and the database open
+operation will be aborted. A database without a valid manifest cannot operate.
+
+#### Handling SST Corruption in Cache Files
+If SlateDB detects corruption in a cache file, it should invalidate the cache
+file and remove the corrupt file. SlateDB will not invoke `DBOptions.on_corruption`
+as there is no action for the operator to take.
+
+
+#### Remaining Available during Corruption
+The rationale for keeping the database available during corruption is to
+provide options to operators and developers building complex systems on top of
+SlateDB.
+
+For instance:
+- If corruption is detected during MemTable flush, it might be desirable for
+  the developer to flush any outstanding buffers to the WAL before closing the
+  database for repair, thereby avoiding data loss.
+- If corruption is detected in the leveled SSTs, it might be desirable for the
+  developer to read uncorrupted keys from the existing SSTs into a new database
+  before fetching the corrupted or missing keys from a replica. This approach
+  reduces the time to recovery.
+- If corruption is detected, the operator may wish to remain available for
+  reads while corruption repair is taking place.
+
+However, leaving the database available for compaction and MemTable flush is a
+dangerous edge case that future patches to SlateDB might not account for.
+Therefore, this aspect should be thoroughly tested to avoid regression.
+
+Operators implementing a logical system on top of SlateDB should utilize the
+`DBOptions.on_corruption` callback to notify other SlateDB read instances of
+the corruption and take appropriate action. Operators may also use the
+`DBOptions.on_corruption` callback to signal the calling system to perform a
+safe database shutdown.
+
+#### Corruption Inspection and Repair
+To support the repair of SSTs, I propose introducing two functions:
+`sst_inspect()` and `sst_repair()`. These functions will provide developers with
+tools to inspect the current contents of a specific SST, and repair a specific
+SST that has been identified as corrupt.
+
+To support repair of the WAL, I propose introducing two functions:
+`wal_inspect()` and `wal_repair()`. These functions will allow developers to
+inspect the current contents of the WAL, including corrupted entries (if possible),
+and to repair the WAL to the best of its ability through skipping corrupted entries 
+or truncation of the WAL.
+
+To support the repair of the Manifest, I propose introducing two functions:
+`manifest_inspect()` and `manifest_repair()`. These functions will enable
```

Comment:
I think this is actually quite challenging, because there is currently no way to reconstruct what the actual sorted runs are from the list of SST files. For the manifest the best we can do is probably to do one of:
1. support reverting the db back to an earlier good manifest
2. allow the user to explicitly construct a manifest that's passed in to repair (a user intimately familiar with slatedb may be able to reconstruct this from log entries, for example)